src/platform/common/variables/customEnvironmentVariablesProvider.node.ts

```diff
@@ -80,19 +91,37 @@ export class CustomEnvironmentVariablesProvider implements ICustomEnvironmentVar
         });
     }
     public createFileWatcher(envFile: string, workspaceFolderUri?: Uri) {
-        if (this.fileWatchers.has(envFile)) {
+        const key = this.getCacheKey(workspaceFolderUri);
+        if (this.fileWatchers.has(key)) {
             return;
         }
-        const envFileWatcher = this.workspaceService.createFileSystemWatcher(envFile);
-        this.fileWatchers.set(envFile, envFileWatcher);
+        const pattern = new RelativePattern(Uri.file(path.dirname(envFile)), path.basename(envFile));
+        const envFileWatcher = this.workspaceService.createFileSystemWatcher(pattern);
+        this.fileWatchers.set(key, envFileWatcher);
         if (envFileWatcher) {
-            this.disposables.push(envFileWatcher.onDidChange(() => this.onEnvironmentFileChanged(workspaceFolderUri)));
-            this.disposables.push(envFileWatcher.onDidCreate(() => this.onEnvironmentFileCreated(workspaceFolderUri)));
-            this.disposables.push(envFileWatcher.onDidDelete(() => this.onEnvironmentFileChanged(workspaceFolderUri)));
+            envFileWatcher.onDidChange(() => this.onEnvironmentFileChanged(workspaceFolderUri), this, this.disposables);
+            envFileWatcher.onDidCreate(() => this.onEnvironmentFileCreated(workspaceFolderUri), this, this.disposables);
+            envFileWatcher.onDidDelete(() => this.onEnvironmentFileChanged(workspaceFolderUri), this, this.disposables);
         }
     }
-    private async _getEnvironmentVariables(resource?: Uri): Promise<EnvironmentVariables> {
-        let customEnvVars = await this.getCustomEnvironmentVariables(resource);
+    private getEnvFile(workspaceFolderUri: Uri, purpose?: 'RunPythonCode') {
```

Comment:
When do we not want to 'runpythoncode'? I feel like if the opposite condition was defined as another string, it would make the use of this more obvious.